feature: auto backflush

niklasarnitz opened the pull request.

niklasarnitz committed:
feature: auto backflush

niklasarnitz commented on src/hardware/state-handlers/GroupHeadHandler.h:
```diff
@@ -24,12 +24,35 @@ class GroupHeadStateHandler : public GenericStateHandler
+        {
+            startAutoBackflush();
+            return;
+        }
```
add that condition

niklasarnitz commented on src/hardware/state-handlers/GroupHeadHandler.h:
```diff
@@ -102,6 +125,62 @@ class GroupHeadStateHandler : public GenericStateHandler
+                }
+            }
+        }
+    }
```
@copilot open a new pull request to apply changes based on [this feedback](https://github.com/niklasarnitz/brewpilot/pull/8#r2513738082)

benasdfasdf commented on src/hardware/handlers/GroupHeadButtonHandler.h:
```diff
@@ -48,6 +48,15 @@ class GroupHeadButtonHandler
+        // Check for auto backflush combo: left single + right double simultaneously held
+        if ((leftSingleDoseEvent == ButtonEventType::BUTTON_HELD) &&
+            (rightDoubleDoseEvent == ButtonEventType::BUTTON_HELD))
+        {
```
Glaub ehrlich gesagt nicht, dass das triggert. Die Events schicken wir immer nur einmal, oder? Dann müsste man jetzt EXAKT gleichzeitig anfangen, die Knöpfe zu drücken, dass das "held" event im gleichen Moment kommt.

Vermutlich müssen wir ins ButtonEventType noch den aktuellen Zustand mit aufnehmen (bool currentlyPressed or sth).
Dann kann man hier prüfen, wenn von einem der beiden ein held (oder auch nur pushed) kommt, ob der andere auch gedrückt ist. Ich würd vllt sogar auf Pressed gehen?
Ich schau mir das mal mit Maschine an, wenn ich daheim bin, wie das am besten funktioniert.

benasdfasdf commented on src/hardware/state-handlers/GroupHeadHandler.h:
```diff
@@ -102,6 +125,62 @@ class GroupHeadStateHandler : public GenericStateHandler
+                              groupNumber, (autoBackflushCycle / 2) + 1);
+                *isExtracting = false;
+                autoBackflushCycle++;
+                autoBackflushStartTime = millis();
```
Würd ich hinters if else ziehen, weil das ja bis auf beim Ende immer benötigt wird - und am Ende juckts niemand, da bleibt da ja eh stuss drin stehen

benasdfasdf commented on src/hardware/state-handlers/StateHandler.h:
```diff
@@ -23,9 +23,9 @@ class StateHandler : public GenericStateHandler
+                               &buttonEvent->groupOne, volumetricsHelper, &state->isInProgrammingMode, 1),
           groupTwoStateHandler(&state->groupTwoIsExtracting,
-                               &buttonEvent->groupTwo, volumetricsHelper, &state->isInProgrammingMode),
+                               &buttonEvent->groupTwo, volumetricsHelper, &state->isInProgrammingMode, 2),
```
Oh I think this has to be rebased on master :sweat_smile:

benasdfasdf reviewed (commented):
yeee

niklasarnitz commented on src/hardware/state-handlers/GroupHeadHandler.h:
```diff
@@ -24,12 +24,35 @@ class GroupHeadStateHandler : public GenericStateHandler
+    unsigned long autoBackflushStartTime = 0;
+    int groupNumber;
+    static const int EXTRACT_DURATION_MS = 10000; // 10 seconds
+    static const int PAUSE_DURATION_MS = 5000;    // 5 seconds
```
yep, seh ich ähnlich.
Die Bluetooth Geschichte wird aber sowieso nochmal ein etwas weiterer Umbau - genau wie die Homeassistant / MQTT geschichte.

Das sollte aber alles auf dem Zweiten core laufen und nur über eine msgqueue miteinander kommunizieren und den state teilen.

benasdfasdf committed:
auto backfush works

it starts as soon as both buttons are pressed and it ends after last extraction instead of awaiting another pause

benasdfasdf commented on src/hardware/handlers/ButtonHandler.h:
```diff
@@ -17,10 +17,16 @@ enum class ButtonEventType
+{
+    ButtonEventType event = ButtonEventType::BUTTON_NONE;
+    bool currentState = false;
+};
```
@niklasarnitz FYI: Neuer Typ für die Rückgabe, der auch den State enthält, nicht nur das event.

benasdfasdf commented on src/hardware/handlers/GroupHeadButtonHandler.h:
```diff
@@ -42,39 +42,48 @@ class GroupHeadButtonHandler
+
+        // Check for auto backflush combo: left single + right double simultaneously held
+        if (((leftSingleDoseState.event == ButtonEventType::BUTTON_PRESSED) && rightDoubleDoseState.currentState) ||
+            ((rightDoubleDoseState.event == ButtonEventType::BUTTON_PRESSED) && leftSingleDoseState.currentState))
```
FYI: Neue Bedingung: Sobald beide gedrückt sind, kann man quasi wieder loslassen. Der erste Button startet seine nomrale extraktion erstmal. Wenn dann auch der zweite gedrückt wird greift diese Bedingung und die Extraktion läuft übers Rückspülen einfach weiter.

benasdfasdf reviewed (commented):
Habs mal zum lauffähig gemacht, tut auch fleißig Dinge. Sowas wie das abbrechen sollte man aber noch bauen.

benasdfasdf commented on src/hardware/state-handlers/GroupHeadHandler.h:
```diff
@@ -102,6 +125,62 @@ class GroupHeadStateHandler : public GenericStateHandler
+        {
+            if (elapsed >= PAUSE_DURATION_MS)
+            {
+                int completedCycles = (autoBackflushCycle / 2) + 1;
```
nehm ich zurück, passt schon.

niklasarnitz commented:
Sieht gut aus - nur das abbrechen muss eig mit rein. 

temporär ginge das ja aber indem man die maschine rebooted.

niklasarnitz commented on src/hardware/state-handlers/GroupHeadHandler.h:
```diff
@@ -104,6 +126,62 @@ class GroupHeadStateHandler : public GenericStateHandler
+        Serial.printf("GroupHeadStateHandler %d: Starting auto backflush sequence\n", groupNumber);
+        isAutoBackflushing = true;
+        autoBackflushCycle = 0;
+        autoBackflushStartTime = millis();
```
this too

benasdfasdf commented on src/hardware/handlers/GroupHeadButtonHandler.h:
```diff
@@ -42,39 +42,48 @@ class GroupHeadButtonHandler
+        {
+            *event = GroupHeadButtonEvent::AUTO_BACKFLUSH;
+            Serial.printf("GroupHeadButtonHandler %d: auto backflush combo detected\n", groupNumber);
+            return;
```
Ich glaub der lügt. Der setzt das ja nur aufs flush event, wenn einer der buttons frisch gedrückt wird (pressed event) und der andere gedrückt ist. Beim halten wird das also nicht nochmal gesetzt.
Am handy find ichs zu unübersichtlich, das ganze zu lesen, aber ich hoffe mal im nächsten cycle würde dann none hier rauskommen.

benasdfasdf replied:
Yep, sieht mir gut aus wie es ist.

benasdfasdf committed:
review comments: can cancel auto flushing

cancel auto flushing with continuous button, don't enable auto flushing when in programming mode, ignore pulses in auto flushing mode, reset pulse counter and target when auto flushing
